server/filestore.go

```diff
@@ -5129,6 +5130,26 @@ func (o *consumerFileStore) encryptState(buf []byte) []byte {
 	return o.aek.Seal(nonce, nonce, buf, nil)
 }
 
+// Used to limit number of disk IO calls in flight since they could all be blocking an OS thread.
+// https://github.com/nats-io/nats-server/issues/2742
+var dios chan struct{}
+
+// Used to setup our simplistic counting semaphore using buffered channels.
+// golang.org's semaphore seemed a bit heavy.
+func init() {
+	// Minimum for blocking disk IO calls.
+	const minNIO = 4
+	nIO := runtime.GOMAXPROCS(0)
```

Comment:
I think we recommend a value of around 4 for this dont we? Is that really enough under heavy load? 

I am not sure how the raft layer behaves exactly but this will presumably not slow down handling of the general keep alive of raft membership right? But could slow down significantly committing data there I presume. Would that increase leader elections under very stressed situations?

Reply:
This is usually set to num cores/cpus, so on my machine its 20 (hyper-threads), but on smaller machines we can use up to 4 instead of setting to 1 or 2. I played with increasing this in some tests but makes no difference since the machine can only be really doing N things in parallel.

This is only effecting the consumer stores, not stream stores which is what is used for nrg/raft layer.

Under a loaded test I had yesterday, performance was much better when lots of ephemeral consumers were trying to do those things all at the same time, and of course the crazy number of threads took its toll on the OS. Did both darwin and linux, so feel pretty good after spending yesterday on it perf wise.

Reply:
Yes, I know defaults but we often say nats server works best on 4ish cores and recommend setting this to 4, was thinking about that